File: rest-service/src/main/java/mb/api/model/reports/ReportResponse.java
package mb.api.model.reports;

import java.util.ArrayList;
import java.util.List;

import com.fasterxml.jackson.annotation.JsonGetter;
import com.fasterxml.jackson.annotation.JsonInclude;
import mb.api.model.io.JsonKeys;
import mb.lib.model.HashID;
import mb.lib.model.JobStatus;

@JsonInclude(JsonInclude.Include.NON_DEFAULT)
public class ReportResponse
{
  private final HashID        jobID;
  private final HashID        reportID;
  private final ReportRequest config;
  private final JobStatus     status;
  private final List<String>  files;
  private final String        description;

  public ReportResponse(
    HashID jobID,
    HashID reportID,
    ReportRequest config,
    JobStatus status,
    String description
  ) {
    this.jobID       = jobID;
    this.reportID    = reportID;
    this.config      = config;
    this.status      = status;
    this.files       = new ArrayList<>();
    this.description = description;
  }

  @JsonGetter(JsonKeys.JobID)
  public HashID getJobID() {
    return jobID;
  }

  @JsonGetter(JsonKeys.ReportID)
  public HashID getReportID() {
    return reportID;
  }

  @JsonGetter(JsonKeys.Config)
  public ReportRequest getConfig() {
    return config;
  }

  @JsonGetter(JsonKeys.Status)
  public JobStatus getStatus() {
    return status;
  }

  @JsonGetter(JsonKeys.Files)
  public List<String> getFiles() {
    return files;
  }

  @JsonGetter(JsonKeys.Description)
  public String getDescription() {
    return description;
  }
}

File: rest-service/src/main/kotlin/mb/api/model/reports/ReportResponse.kt
package mb.api.model.reports

import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.annotation.JsonValue
import com.fasterxml.jackson.databind.node.ObjectNode
import mb.api.model.io.JsonKeys
import mb.lib.model.JobStatus
import org.veupathdb.lib.hash_id.HashID
import org.veupathdb.lib.jackson.Json

data class ReportResponse(
  @JsonProperty(JsonKeys.JobID)       val jobID:       HashID,
  @JsonProperty(JsonKeys.ReportID)    val reportID:    HashID,
  @JsonProperty(JsonKeys.Config)      val config:      ReportRequest,
  @JsonProperty(JsonKeys.Status)      val status:      JobStatus,
  @JsonProperty(JsonKeys.Description) val description: String? = null,
) {
  @JsonProperty(JsonKeys.Files)
  val files: MutableList<String> = ArrayList()

  @JsonValue
  fun toJson(): ObjectNode =
    Json.new {
      put(JsonKeys.JobID, jobID.string)
      put(JsonKeys.ReportID, reportID.string)
      set<ObjectNode>(JsonKeys.Config, config.toJson())
      put(JsonKeys.Status, status.value)
      description?.let { put(JsonKeys.Description, it) }

      if (files.isNotEmpty())
        putPOJO(JsonKeys.Files, files)
    }
}